kroxylicious-filters/kroxylicious-sasl-inspection/src/main/java/io/kroxylicious/filters/sasl/inspection/SaslInspection.java

```diff
@@ -0,0 +1,86 @@
+import io.kroxylicious.proxy.plugin.PluginConfigurationException;
+import io.kroxylicious.proxy.tag.VisibleForTesting;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+/**
+ * Factory for {@link SaslInspectionFilter}.
+ */
+@Plugin(configType = Config.class)
+public class SaslInspection implements FilterFactory<Config, Void> {
+
+    private Map<String, SaslObserverFactory> observerFactoryMap;
+
+    @Override
+    public Void initialize(FilterFactoryContext context,
+                           @Nullable Config config)
+            throws PluginConfigurationException {
+        observerFactoryMap = buildEnabledObserverFactoryMap(context, config);
+        return null;
+    }
+
+    @Override
+    public Filter createFilter(FilterFactoryContext context, Void unused) {
+        return new SaslInspectionFilter(observerFactoryMap);
+    }
+
+    private Map<String, SaslObserverFactory> buildEnabledObserverFactoryMap(FilterFactoryContext context, @Nullable Config config) {
+        var allNames = context.pluginImplementationNames(SaslObserverFactory.class);
+        var allMap = allNames
+                .stream()
+                .map(instanceName -> context.pluginInstance(SaslObserverFactory.class, instanceName))
+                .collect(Collectors.toMap(SaslObserverFactory::mechanismName, Function.identity(),
+                        (sof1, sof2) -> {
+                            if (sof1.getClass() != sof2.getClass()) {
+                                throw new IllegalStateException(sof1.getClass().getSimpleName() + " and " + sof2.getClass().getSimpleName()
+                                        + " both register the same SASL mechanism name " + sof1.mechanismName());
+                            }
+                            return sof1;
+                        }));
+
+        var secureOnly = allMap.entrySet()
+                .stream().filter(Predicate.not(e -> e.getValue().transmitsCredentialInCleartext()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (x, y) -> y, HashMap::new));
+
+        var enabledMechNames = Optional.ofNullable(config).map(Config::enabledMechanisms).orElse(secureOnly.keySet());
+
+        var unknownMechNames = enabledMechNames.stream().filter(Predicate.not(allMap::containsKey)).collect(Collectors.toSet());
+        if (!unknownMechNames.isEmpty()) {
+            throw new PluginConfigurationException("The following enabled SASL mechanism names are unknown: [" + String.join(",", unknownMechNames) + "]");
+        }
+
+        Map<String, SaslObserverFactory> onlyEnabled = new HashMap<>(allMap);
+        onlyEnabled.keySet().retainAll(enabledMechNames);
+
+        if (onlyEnabled.isEmpty()) {
+            throw new PluginConfigurationException("The SaslObserver requires at at least one enabled SaslObserver implementation. "
+                    + "Discovered implementations names: [" + String.join(",", allNames) + "]");
+        }
```

Comment:
I'm wondering if we want more explicit logging/errors for the auto-configuration case.

If we are auto-configuring mechanisms and no appropriate mechanisms were found, we could fail with an error like: "no SASL mechanisms were automatically selected for inspection from the discovered mechanism names ['PLAIN'], please specify the enabled SASL mechanisms in your filter configuration using field 'enabledMechanisms'".

Also wonder if we should add an info log to describe the enabled mechanisms on the happy path, to expose what the auto-configuration decided. Or maybe explicitly log if we go down the auto-config path.

Reply:
and should both errors be mechanism oriented, it might be a bit jarring for the user to see an error about SaslObserver implementation names when the filter configuration is in terms of mechanisms. And on the other hand information about the SaslObserver could be useful for power users, maybe we should debug log some detail about all the discovered SaslObservers.